packages/lib/services/e2ee/types.ts

```diff
@@ -25,3 +25,24 @@ export interface RSA {
 	publicKey(rsaKeyPair: RSAKeyPair): string;
 	privateKey(rsaKeyPair: RSAKeyPair): string;
 }
+
+export interface Crypto {
+	getCiphers(): string[];
+	getHashes(): string[];
+	randomBytes(size: number): Promise<CryptoBuffer>;
+	pbkdf2Raw(password: string, salt: CryptoBuffer, iterations: number, keylen: number, digest: Digest): Promise<CryptoBuffer>;
+}
+
+export interface CryptoBuffer extends Uint8Array {
+	toString(encoding?: BufferEncoding, start?: number, end?: number): string;
+}
+
+// From react-native-quick-crypto.HashAlgorithm, but use the hash name style in node:crypto
+export enum Digest {
+	sha1 = 'sha1',
+	sha224 = 'sha224',
+	sha256 = 'sha256',
+	sha384 = 'sha384',
+	sha512 = 'sha512',
+	ripemd160 = 'ripemd160',
```

Comment:
We need to think a bit more about this. So you are saving the digest names as plain string - what was the reasoning for this? As opposed to saving them as plain integer for example (1, 2, 3, 4, etc.)?

Look for example at the `EncryptionMethod` enum - those are all integers, and that's because they are part of each chunk of encrypted data, so they will appear potentially hundreds of time in an encrypted note, and billions of times on a server that will store many notes for many users. So it makes sense to be efficient and use an INT here. It's also more efficient if that info will appear in a column in a database.

How about this Digest info - how will it be used? How will it be stored, and where will it appear?

Reply:
> So you are saving the digest names as plain string - what was the reasoning for this?

The `digest` parameter in both `node:crypto` and `react-native-quick-crypto` accepts a string object. If we make the enum as string we don't need extra conversions.

> How about this Digest info - how will it be used? How will it be stored, and where will it appear?

In my current design it will only be used in a wrapper function. I will keep using `sha512` so this parameter won't be specified in the encrypted data and won't cause extra size overhead.

Reply:
Plus, the `Digest` is only used in `pbkdf2Raw()` which is a bare export of the platform-specific implementation. I think it should be similar to the raw implementation, and the behavior of encryption and the format of ciphertext should be defined in the wrapper function.

Reply:
> The digest parameter in both node:crypto and react-native-quick-crypto accepts a string object. If we make the enum as string we don't need extra conversions.

Mapping an int to a string is not an issue

> Plus, the Digest is only used in pbkdf2Raw() which is a bare export of the platform-specific implementation. I think it should be similar to the raw implementation

Your implementation still requires a mapping from your own type to the lib type so it's already different.

It's still not clear to me whether this info will be stored or not (or even why we need a hashing function), but it's ok let's leave this PR open anyway, and please create the next PR based on this one. If you need help on how to do this please let me know

Reply:
> If you need help on how to do this please let me know.

I guess I only need to rebase the changes for the next PR to this branch, right?

Reply:
> It's still not clear to me whether this info will be stored or not (or even why we need a hashing function)

The hash function is used in PBKDF2 key derivation algorithm.
Let's take the current encryption process in Joplin as an example:
1. When calling `sjcl.json.encrypt()`, it [calls `sjcl.misc.cachedPbkdf2()`](https://github.com/bitwiseshiftleft/sjcl/blob/1.0.8/core/convenience.js#L49) to generate a key from the password or use the key cache.
2. In `sjcl.misc.cachedPbkdf2()`, if the cache is not available, it [calls `sjcl.misc.pbkdf2()`](https://github.com/bitwiseshiftleft/sjcl/blob/1.0.8/core/convenience.js#L325) with no `Prff` parameter to generate the key.
3. When parameter `Prff` is not set, it [will be `sjcl.misc.hmac()`](https://github.com/bitwiseshiftleft/sjcl/blob/1.0.8/core/pbkdf2.js#L37-L39) with no `Hash` parameter, and the default hash [will be `sha256`](https://github.com/bitwiseshiftleft/sjcl/blob/1.0.8/core/hmac.js#L13-L14) here.

The encrypted data of Joplin doesn't save the hash algorithm name because it uses the default hash algorithm when calling `sjcl.misc.pbkdf2()`. For my implementation, I will set the default value to `sha512` in the wrapper function, so we don't need to save the hash algorithm too.